Add picture to personal data

vtran99 opened the pull request:
Request to add the picture to personal data type of registration.
This will help to display picture for registrations in Participant List or Badges.

ThiefMaster commented on indico/modules/events/registration/models/items.py:
```diff
@@ -49,7 +49,7 @@ class PersonalDataType(IndicoIntEnum):
 
     __titles__ = [None, 'Email Address', 'First Name', 'Last Name', 'Affiliation', 'Title', 'Address',
-                  'Phone Number', 'Country', 'Position']
+                  'Phone Number', 'Country', 'Position', 'Picture Id']
```
What's a "Picture Id"? I think this should be just "Picture".

vtran99 replied:
Done.

ThiefMaster commented on indico/modules/events/registration/models/registrations.py:
```diff
@@ -452,6 +452,16 @@ def avatar_url(self):
+    @property
+    def picture_url(self):
+        """Return the url of the picture in personal data."""
+        picture = [d for d in self.data if d.field_data.field.personal_data_type is not None
```
why `is not None` and not check for `PersonalDataType.picture`?

ThiefMaster commented on indico/modules/events/registration/models/registrations.py:
```diff
@@ -452,6 +452,16 @@ def avatar_url(self):
+        picture = [d for d in self.data if d.field_data.field.personal_data_type is not None
+                   and d.field_data.field.input_type == 'picture']
+        if picture and picture[0].filename:
+            return url_for('event_registration.registration_picture', self,
```
I don't think you can reliably define a property for this. It heavily depends on the context which URL to use (for example if you're viewing your own registration you use the `self.locator.registrant`, but as a manger you wouldn't do that.

I recommend you don't add such a property now, and when you add something for the participant list you create a new endpoint (needed anyway since it'll have to check if the current user can see that person in the participant list and the picture is supposed to be shown there) and use that then (whether this property then makes sense or not is to be seen).

vtran99 replied:
Property removed.

tomasr8 commented:
Does this need a new `RegistrationPDPlaceholder`?

vtran99 commented:
> Does this need a new `RegistrationPDPlaceholder`?

Well, to be complete, it's not only a placeholder in designer, but should also be in:
- registration list
- participant list
- ...
and the picture could be rendered as url, binary, whatever form... so I suggest to wait and add that part following the feature needed.

ThiefMaster commented:
I think adding a placeholder to include it on badge would make sense in this PR. I guess that's something you guys need anyway?

vtran99 commented:
> I think adding a placeholder to include it on badge would make sense in this PR. I guess that's something you guys need anyway?

Done.

tomasr8 commented:
> > I think adding a placeholder to include it on badge would make sense in this PR. I guess that's something you guys need anyway?
> 
> Done.

There are still some lint/test issues, could you fix them?

vtran99 commented:
> > > I think adding a placeholder to include it on badge would make sense in this PR. I guess that's something you guys need anyway?
> > 
> > 
> > Done.
> 
> There are still some lint/test issues, could you fix them?

Done.

ThiefMaster commented on indico/modules/designer/placeholders_test.py:
```diff
@@ -124,4 +124,4 @@ def test_render_image_placeholders(dummy_event, dummy_reg, dummy_designer_image_
         kwargs = _get_render_kwargs(ph, dummy_event, dummy_person, dummy_reg, dummy_item)
         res = ph.render(**kwargs)
-        assert isinstance(res, Image.Image)
+        assert (isinstance(res, Image.Image) or res is None)
```
```suggestion
        assert isinstance(res, Image.Image) or res is None
```

ThiefMaster commented on indico/modules/events/registration/models/registrations.py:
```diff
@@ -595,7 +595,7 @@ def get_personal_data(self):
             field = data.field_data.field
-            if field.personal_data_type is not None and data.data:
+            if field.personal_data_type is not None and (data.data or data.filename):
                 personal_data[field.personal_data_type.name] = data.friendly_data
```
Maybe in case of a picture field we should add a reference to the data object itself (since it contains `.open()` etc? That way you could use `get_personal_data()` e.g. when rendering the placeholder instead of having to manually loop over the data.

vtran99 replied:
If we use the new property, in render of PicturePlaceholder =>
picture_data = registration.picture_data

ThiefMaster replied:
Maybe not adding it to `get_personal_data` but just having `get_personal_picture` might be better then? (the filename in the personal data is pretty much useless)

vtran99 replied:
"the filename in the personal data is pretty much useless": I know :-)
it's just to mean there is a picture in the registration.

I don't get what you meant:
 add a method 'get_personal_picture' ? but what to return in picture for get_personal_data ?

ThiefMaster replied:
you're right, since `open()` is meant to be used in a `with` block that mathod would not be particularly useful. Maybe `get_personal_data_picture` that returns the actual RegistrationData for that field, sine this is where you have e.g. the `open` and `send` methods

vtran99 replied:
> get_personal_data_picture that returns the actual RegistrationData for that field
=> Ok,
but what would you like to get for picture in "get_personal_data" ?
(currently I just return the useless filename in personal data)

ThiefMaster replied:
I would be completely OK with skipping it altogether, or alternatively including the RegistrationData object there directly.

Actually, after looking at the code: Building the full personal data dict exactly as (in)expensive as just getting the RegistrationData for the picture, since you're probably going to iterate over all the personal data fields in either case. So maybe the "standalone" method is not necessary, and if you need the picture regdata you just use `.get_personal_data()['picture']`?

ThiefMaster commented on indico/modules/designer/placeholders.py:
```diff
@@ -361,6 +362,20 @@ class RegistrationPhonePlaceholder(RegistrationPDPlaceholder):
+        picture_data = [d for d in registration.data
+                        if d.field_data.field.personal_data_type == PersonalDataType.picture]
+        if picture_data and picture_data[0].filename:
+            buf = picture_data[0].open()
+            return Image.open(buf)
```
- I think `.open()` whould be used with a `with` block.
- What happens when rendering the template for a registration that has no picture (ie where this method here returns `None`)?

ThiefMaster replied:
Yes of course. What I meant is when the render method returns None

ThiefMaster commented on indico/migrations/versions/20240130_1531_17996ef18cb9_update_constraint_valid_enum_personal_data_type.py:
```diff
@@ -0,0 +1,29 @@
+"""Update constraint valid_enum_personal_data_type
```
```suggestion
"""Add picture personal data type for registrations
```

and I'd rename the file to `20240130_1531_17996ef18cb9_add_picture_personal_data_type_registrations.py`

Context: When looking for an alembic revision I usually search in the filenames for something I know it touched. In this case it'd be "registration" and "picture".

Also, that title is nicer IMHO ;)

vtran99 commented:
I removed the picture from get_personal_data and added a stand-alone get_personal_data_picture, which will be used anytime the picture needs to be retrieved (to avoid having to build the whole personal data dict).

ThiefMaster commented on indico/modules/events/registration/badges.py:
```diff
@@ -134,8 +134,9 @@ def _draw_badge(self, canvas, person, template, tpl_data, pos_x, pos_y):
                 item_data.update(update)
-            self._draw_item(canvas, item_data['item'], tpl_data, item_data['text'], item_data['pos_x'],
-                            item_data['pos_y'])
+            if item_data['text'] is not None:
```
FYI this is what my initial comment about "what if render() returns None" was about:

![image](https://github.com/indico/indico/assets/179599/49f9726f-a2d3-46d7-9f9f-7e6bc661313b)

Before adding this fix, the badge got a literal `None` printed on it.

ThiefMaster reviewed (commented):
I fixed a bunch of bugs, I think this is fine now. But have a quick look over my fixes.

ThiefMaster commented:
One thing that might be worth fixing here is the fact that the aspect ratio is not honored when the picture is rendered on a badge. Cropping may lose some important data, so maybe it should be resized so it fits in the size of the badge placeholder (and then just has some more blank space around it when the aspect ratios don't match)

ThiefMaster commented:
Squeezing the image in such a way feels like a bug to me that should be fixed before merging it upstream...

vtran99 commented:
Ok for you if I use the default 'preserveAspectRatio' of the library ?
It will center the picture in the frame, the blank space could be around width/height depending of the picture format...

ThiefMaster commented:
Not sure if that would work since you can still upload any aspect ratio (seems reasonable, the user may upload a landscape picture but then crop it to only have the relevant part with their picture which is not landscape anymore), and the badge template is completely unrelated to that.

Maybe the regform picture field needs a setting to enforce a certain aspect ratio in case  this is needed (actually I'm surprised you do not have this requirement at the UN - the last time I got a UN badge my photo was actually squished which did not look very flattering; some people might even be annoyed/offended by that ;))

vtran99 commented:
I did a test with the case you mentioned: upload a landscape pict, crop to get just the face part, the rendering is correct with preserveAspectRatio, as the picture saved is the new one created from the crop, with the format of the crop and not the landscape format anymore.
To resume, the picture rendered in the draw_image with preserveAspectRatio will match the ratio of picture saved (cropped or not) of the upload.

ThiefMaster commented:
But how will that help when it gets squeezed in the size of the placeholder in the badge template? If that placeholder is square and the image is e.g. in 2:3 format, it will still get squished, won't it?

vtran99 commented:
It will be rendered in 2:3 format, with blank to fill the placeholder.
And that's correct, if user uploads a picture with 2:3 format, will see a 2:3 format !

vtran99 commented:
<img width="548" alt="badge-flag" src="https://github.com/indico/indico/assets/11231950/4b8279e1-4dea-4873-81b6-9dd8183ec727">

vtran99 commented:
The placeholder is a square, and the picture uploaded in 2:3, the rendering is 2:3.

vtran99 commented:
picture vertical when uploaded, badge rendered with same square placeholder:
<img width="548" alt="Screenshot 2024-02-12 at 16 02 38" src="https://github.com/indico/indico/assets/11231950/90133060-25b7-4330-b903-16cac1c9a923">

ThiefMaster commented:
looks good :)

vtran99 commented:
How about my suggestion to change the _reglist.html ?
just modify the td condition instead of adding a search_value 
`<td class="i-table">{{ data[item.id].friendly_data if item.id in data and data[item.id].friendly_data else '' }}</td>`

ThiefMaster commented:
I don't see where you made that... (or is it new?).

Anyway, this is generic and not specific to the picture field, so I'd prefer to use the same logic we use in other places in the template.

vtran99 commented:
All done on my side related to your review.

ThiefMaster commented:
Oh now I understand after seeing your commit. I did not expect camelCase in the Python code so I assumed this is a prop you pass somewhere on the image cropping UI of the regform!
